agent/hcp/telemetry/reporter.go

```diff
@@ -0,0 +1,150 @@
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/consul/lib"
+	"github.com/hashicorp/go-hclog"
+)
+
+const (
+	defaultStreamTimeout  = time.Minute
+	defaultReportInterval = 5 * time.Second
+	defaultBatchInterval  = 10 * time.Second
+)
+
+type ReporterConfig struct {
+	StreamTimeout  time.Duration
+	ReportInterval time.Duration
+	BatchInterval  time.Duration
+
+	Logger hclog.Logger
+
+	Gatherer lib.MetricsHandler
+	Exporter Exporter
+}
+
+func DefaultConfig() *ReporterConfig {
+	return &ReporterConfig{
+		StreamTimeout:  defaultStreamTimeout,
+		ReportInterval: defaultReportInterval,
+		BatchInterval:  defaultBatchInterval,
+	}
+}
+
+type Reporter struct {
+	cfg ReporterConfig
+
+	batchedMetrics       map[time.Time]*metrics.IntervalMetrics
+	flushCh              chan struct{}
+	lastIntervalExported time.Time
+
+	// Only used for test purposes to know when flush has occurred.
+	testFlushCh chan struct{}
+}
+
+func NewReporter(cfg *ReporterConfig) (*Reporter, error) {
+	if cfg.Exporter == nil || cfg.Gatherer == nil || cfg.Logger == nil {
```

Comment:
Safety guard, a client / logger are needed. If this package is used outside of our use case, this ensures the config object is configured well.